fix xmlrpc timeout using monotonic clock

flixr opened the pull request:
in XmlRpcDispatch::work the _endTime check was using system time instead of monotonic time,
so when time was set back it didn't process all events..

flixr committed:
fix xmlrpc timeout using monotonic clock

in XmlRpcDispatch::work the _endTime check was using system time instead of monotonic time,
so when time was set back it didn't process all events..

mikepurvis commented:
LGTM, but this should be confirmed to work correctly OS X and FreeBSD.

flixr commented:
Good point.
According to https://www.unix.com/man-page/FreeBSD/2/clock_gettime/ it's available on FreeBSD, but apparently not on OSX: https://stackoverflow.com/questions/5167269/clock-gettime-alternative-in-mac-os-x

mikepurvis commented:
Reference implementation:

https://github.com/wjwwood/serial/blob/827c4a784dd4fdd35dc391f37ef152eab7c9c9b2/src/impl/unix.cc#L81-L97

flixr committed:
replace clock_gettime for OSX

dirk-thomas commented:
The same code already exists in [rostime](https://github.com/ros/roscpp_core/blob/64da30f22720ded414c4df37e9e0be41993c277e/rostime/src/time.cpp#L185). It should either be reused or at last be copied with a comment from where the code is coming and the same PP conditions should be used.

flixr committed:
xmlrpcpp: copy code from ros_steadytime for getTime

dirk-thomas commented:
Instead of copy-n-pasting a non-trivial code block like this wouldn't it be better to expose `ros_steadytime` in a header in `rostime` and call the function here?

flixr commented:
I thought it would be better to not add this dependency here, but I could do that if this is preferred...

flixr commented:
@dirk-thomas is it ok to leave it like this, or should we really add rostime as a dependency to xmlrpcpp?

flixr commented:
ping @dirk-thomas

dirk-thomas commented:
I think the risk is pretty high that when the implementation in `rostime` is updated / fixed / modified in the future this copied code isn't. Therefore I think it is worth to expose the function in `rostime` in a header and update this patch to use the new function (and declare a version dependency on `rostime` with the next patch release.

flixr committed:
use ros_steadytime

which depends on https://github.com/ros/roscpp_core/pull/73
and assumes that it will be available with rostime 0.6.9

dirk-thomas committed:
fold build and run depend into depend tag to avoid version to be repeated

dirk-thomas commented:
I will have to release `roscpp_core` into Lunar in order for the PR job to pass... (ros/rosdistro#16898)

dirk-thomas commented:
@ros-pull-request-builder retest this please

dirk-thomas commented on utilities/xmlrpcpp/src/XmlRpcDispatch.cpp:
```diff
@@ -222,11 +224,10 @@ XmlRpcDispatch::getTime()
 
-  gettimeofday(&tv, &tz);
-  return (tv.tv_sec + tv.tv_usec / 1000000.0);
+  ros_steadytime(sec, nsec);
```
This might need the `ros::` namespace?

dirk-thomas committed:
add missing namespace to ros_steadytime

dirk-thomas committed:
add missing target_link_libraries

dirk-thomas commented:
I tried to fix this "on-the-fly" using the GitHub web interface which wasn't really successful. Can you please update the PR to make it pass CI. Thanks.

flixr committed:
add missing lib to tests

flixr commented:
Done.
Sorry about the missing things earlier... I didn't have machine handy to test it and wrongly thought it would be trivial enough to do without tests..

dirk-thomas commented:
No problem. Thank you for addressing it quickly.
